rust/connlib/tunnel/src/control_protocol/client.rs

```diff
@@ -84,118 +92,114 @@ where
+            .role_state
+            .lock()
+            .gateway_preshared_keys
+            .get(&gateway_id)
+            .ok_or(Error::ControlProtocolError)?
+            .clone();
+        let index = self.next_index();
 
-                if let Some(conn) = tunnel.peer_connections.lock().get(&gateway_id) {
-                    conn.on_peer_connection_state_change(on_peer_connection_state_change_handler(
-                        gateway_id,
-                        tunnel.stop_peer_command_sender.clone(),
-                    ));
-                }
+        let peer_config = self
+            .role_state
+            .lock()
+            .create_peer_config_for_new_connection(resource_id, gateway_id, p_key)?;
 
-                tokio::spawn(tunnel.start_peer_handler(peer, d));
+        let peer = Arc::new(Peer::new(
+            self.private_key.clone(),
+            index,
+            peer_config.clone(),
+            gateway_id,
+            None,
+            self.rate_limiter.clone(),
+        ));
 
-                tunnel
-                    .role_state
-                    .lock()
-                    .awaiting_connection
-                    .remove(&resource_id);
-            })
-        }));
+        let ep = ice
+            .new_endpoint(Box::new(|_| true))
+            .await
+            .ok_or(Error::ControlProtocolError)?;
 
-        let offer = peer_connection.create_offer(None).await?;
-        peer_connection.set_local_description(offer.clone()).await?;
+        let (peer_sender, peer_receiver) = tokio::sync::mpsc::channel(PEER_QUEUE_SIZE);
 
-        Ok(Request::NewConnection(RequestConnection {
-            resource_id,
+        tokio::spawn({
+            let ep = ep.clone();
+            async move {
+                peer_handler::handle_packet(ep, peer_receiver).await;
+            }
+        });
+
+        ice.on_connection_state_change(on_peer_connection_state_change_handler(
             gateway_id,
-            client_preshared_key: Secret::new(Key(preshared_key.to_bytes())),
-            client_rtc_session_description: offer,
-        }))
+            self.stop_peer_command_sender.clone(),
+        ));
+
+        self.role_state
+            .lock()
+            .peer_queue
+            .insert(gateway_id, peer_sender);
+
+        {
+            let mut peers_by_ip = self.peers_by_ip.write();
+
+            for ip in peer_config.ips {
+                peers_by_ip.insert(
+                    ip,
+                    ConnectedPeer {
+                        inner: peer.clone(),
+                        channel: ep.clone(),
+                    },
+                );
+            }
+
+            self.role_state
+                .lock()
+                .gateway_awaiting_connection
+                .remove(&gateway_id);
+        }
+
+        tokio::spawn(self.start_peer_handler(peer, ep));
+
+        self.role_state
+            .lock()
+            .awaiting_connection
+            .remove(&resource_id);
```

Comment:
Can we combine these operations on `role_state` into a function perhaps? Possibly even with the ones further above?